org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java

```diff
@@ -39,13 +45,43 @@ public boolean forgetLastBuildState(Iterable<IProject> toUpdate, Map<String, Str
 				XtextBuilder builder = BuildManagerAccess.findBuilder(project);
 				if (builder != null) {
 					builder.forgetLastBuiltState();
+					touchProject(project);
 				}
 			}
 			return true;
 		}
 		return false;
 	}
 
+	/**
+	 * Touch a given project such that the auto-build knows it should be rebuild.
+	 *
+	 * @since 2.26
+	 */
+	protected void touchProject(IProject project) {
+		if (!project.getWorkspace().isTreeLocked()) {
```

Comment:
Thank you for this PR, it also fixed my issue #1676, but I recognized in the builds some exceptions due to a missing check in case for closed projects. When I add project.isAccessible() to that condition and before line 64 in the touch from the job, then our builds are green

Reply:
@rytina could you provide a pr

Reply:
or create an addtional issue? i also wonder how you can bypass the
BuildManagerAccess.findBuilder calling isAccessible in line 62

Reply:
https://github.com/eclipse/xtext-eclipse/pull/1678

Reply:
@szarnekow  any idea if the strange messages in https://github.com/itemis/xtext-reference-projects/runs/2671804040?check_suite_focus=true could be related?

Reply:
(deadlock)

Reply:
```
!MESSAGE Deadlock detected. All locks owned by thread Worker-6: Schedule Xtext recovery build on start-up will be suspended.
!STACK 0
java.lang.IllegalStateException
	at org.eclipse.core.internal.jobs.DeadlockDetector.reportDeadlock(DeadlockDetector.java:604)
	at org.eclipse.core.internal.jobs.DeadlockDetector.lockWaitStart(DeadlockDetector.java:408)
	at org.eclipse.core.internal.jobs.LockManager.addLockWaitThread(LockManager.java:147)
	at org.eclipse.core.internal.jobs.OrderedLock.doAcquire(OrderedLock.java:171)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:109)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:85)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState$1.run(AbstractBuilderState.java:83)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2292)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2317)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.load(AbstractBuilderState.java:103)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.ensureLoaded(AbstractBuilderState.java:118)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.getAllResourceDescriptions(AbstractBuilderState.java:139)
	at org.eclipse.xtext.builder.impl.ToBeBuiltComputer.doRemoveProject(ToBeBuiltComputer.java:204)
	at org.eclipse.xtext.builder.impl.ToBeBuiltComputer.updateProject(ToBeBuiltComputer.java:264)
	at org.eclipse.xtext.builder.impl.XtextBuilder.fullBuild(XtextBuilder.java:424)
	at org.eclipse.xtext.builder.impl.XtextBuilder.build(XtextBuilder.java:202)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:846)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:229)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:277)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:330)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:333)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:385)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:406)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:154)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:244)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
!SUBENTRY 1 org.eclipse.core.jobs 4 2 2021-05-26 04:58:22.818
!MESSAGE Thread Worker-5: Building has locks: R/, P/xtend-euler and is waiting for lock OrderedLock (4)
!SUBENTRY 1 org.eclipse.core.jobs 4 2 2021-05-26 04:58:22.820
!MESSAGE Thread Worker-6: Schedule Xtext recovery build on start-up has locks: OrderedLock (4) and is waiting for lock P/xtend-euler
```

Question is: What is OrderedLock(4) :)

Reply:
Might well be that we always need the job since isTreeLocked might give us the wrong result for our context - e.g. if we are running a build from the current thread, a touch might be the wrong thing to happen.

Reply:
will try to run the stuff locally over the weekend.
then i will hopefully be able to provide a complete jstack

Reply:
[x.log](https://github.com/eclipse/xtext-eclipse/files/6561103/x.log)

Reply:
[y.log](https://github.com/eclipse/xtext-eclipse/files/6561186/y.log)

Reply:
problem is reproducing locally.
undoing the change makes test green again.
it seems to "always" run in a job at least also helps.
will try to verify with jenkins/gh actions

Reply:
looks promising https://github.com/itemis/xtext-reference-projects/actions/runs/886423181